Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Utils/MultiSparseVector.h

```diff
@@ -136,6 +142,59 @@ namespace AZ::Render
     template<typename ... Ts>
     void MultiSparseVector<Ts...>::Clear()
     {
+        // Because the memory in the underlying vector is used to store a linked list of the removed items,
+        // a destructor could be called on bogus memory when the vector is cleared or destroyed. To fix this,
+        // through each free slot and default-construct an object there so it can be safely deleted.
+
+        // First create a tuple which only contains the vectors with non-trivial destructors.
+        auto TuplesToReset = [](auto&... dataVectors)
+        {
+            auto AddVectorToTuple = [](auto& dataVector)
+            {
+                using VectorType = AZStd::remove_cvref_t<decltype(dataVector)>;
+                using ValueType = typename VectorType::value_type;
+                if constexpr (!AZStd::is_trivially_destructible_v<ValueType>)
+                {
+                    return AZStd::tuple<VectorType&>(dataVector);
+                }
+                else
+                {
+                    // Exclude the type if it is trivially destructible from the reset logic
+                    return AZStd::tuple<>{};
+                }
+            };
+            return AZStd::tuple_cat(AddVectorToTuple(dataVectors)...);
+        };
+
+        auto tupleVectorsWithNonTrivialDestructors = AZStd::apply(TuplesToReset, m_data);
+        constexpr bool containsNonTrivialDestructibleVector = AZStd::tuple_size_v<decltype(tupleVectorsWithNonTrivialDestructors)> > 0;
+
+        // If there are any vectors with non-trivial destructors, then go through each one and default-initialize
+        // it so it can be safely deleted when the underlying vector's clear() is called.
+        if constexpr (containsNonTrivialDestructibleVector)
```

Comment:
Not having that check leaves it up to the compiler to figure out, which I'd assume it _usually_ does the right thing. However, since this only adds a single constexpr if check on top of what's already necessary, I think it's worth doing to be sure. It may even save the compiler some time so it doesn't have to figure out how to optimize the code itself.

Reply:
Actually not having the check would result in a compile error.
Because the call to access the first vector via `AZStd::get<0>` would fail because the tuple is empty.

Reply:
I think it would still work - you could have a MultiSparseVector of all trivially destructible objects resulting in an empty tuple, but `AZStd::get<0>(m_data)` would be fine. I'm not sure you can even construct a MultiSparseVector without at least one type

Reply:
No, that would result in an out of bounds error: https://godbolt.org/z/98MYWfWo4

Reply:
`m_data` is not 'tupleVectorsWithNonTrivialDestructors' though, it's the original tuple for the MultiSparseVector. It could just be a bunch of pod structs. This would leave 'tupleVectorsWithNonTrivialDestructors' empty, but 'm_data' would still have entries in it.

Reply:
Ok, I see the `AZStd::get<0>` call is on the original tuple, not the sub tuples of types with non-trivial destructors.
So you are correct, it would still compile.